library/std/src/sys/hermit/fs.rs

```diff
@@ -119,75 +165,127 @@ impl FileTimes {
 
 impl FileType {
     pub fn is_dir(&self) -> bool {
-        self.0
+        self.mode == DT_DIR
     }
-
     pub fn is_file(&self) -> bool {
-        self.0
+        self.mode == DT_REG
     }
-
     pub fn is_symlink(&self) -> bool {
-        self.0
+        self.mode == DT_LNK
     }
 }
 
-impl Clone for FileType {
-    fn clone(&self) -> FileType {
-        self.0
+impl fmt::Debug for ReadDir {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // This will only be called from std::fs::ReadDir, which will add a "ReadDir()" frame.
+        // Thus the result will be e g 'ReadDir("/home")'
+        fmt::Debug::fmt(&*self.inner.root, f)
     }
 }
 
-impl Copy for FileType {}
+impl Iterator for ReadDir {
+    type Item = io::Result<DirEntry>;
 
-impl PartialEq for FileType {
-    fn eq(&self, _other: &FileType) -> bool {
-        self.0
-    }
-}
+    fn next(&mut self) -> Option<io::Result<DirEntry>> {
+        if self.end_of_stream {
+            return None;
+        }
 
-impl Eq for FileType {}
+        unsafe {
+            loop {
+                // As of POSIX.1-2017, readdir() is not required to be thread safe; only
+                // readdir_r() is. However, readdir_r() cannot correctly handle platforms
+                // with unlimited or variable NAME_MAX. Many modern platforms guarantee
+                // thread safety for readdir() as long an individual DIR* is not accessed
+                // concurrently, which is sufficient for Rust.
+                let entry_ptr = match abi::readdir(self.inner.dirp.as_raw_fd()) {
+                    abi::DirectoryEntry::Invalid(e) => {
+                        // We either encountered an error, or reached the end. Either way,
+                        // the next call to next() should return None.
+                        self.end_of_stream = true;
+
+                        return Some(Err(Error::from_raw_os_error(e)));
+                    }
+                    abi::DirectoryEntry::Valid(ptr) => {
+                        if ptr.is_null() {
+                            return None;
+                        }
+
+                        ptr
+                    }
+                };
+
+                macro_rules! offset_ptr {
```

Comment:
I'm not sure why this macro exists. Most/all of the use sites could be fine with another construct (either addr_of, or just `(*p).field`), no?

The only place I see something like it being needed is for the name field of the dirent (since it might not be entirely there) but at that point you can just use offset_of and casting, no?